File: 2_implementacion_algoritmos/cpp/counting.cpp

// Algoritmo de ordenamiento - INSERTION SORT

#include <bits/stdc++.h>
#include <sstream>
#include <string>
#include <vector>
#include <iostream>
#include <fstream>

using namespace std;

// imprimir arreglo
void printArray(int arr[], size_t sizeList){    
    for (int i = 0; i < sizeList; i++) {        
        cout<< arr[i]<<" ";		
    }        
}
 // retorna el maximo valor de un arreglo de numeros enteros
int maxValueInArray(int list[], size_t sizeList) {
	int max = 0;
	for (int i = 0; i < sizeList; i++) {
		max = (max >= list[i]) ? max : list[i];
	}
	
	return max;
}

// A = lista no ordenada inicial.
// B = lista ordenada para retornar.
// k = valor maximo de la lista.
// sizeList = tamaño de la lista A
int *counting_sort(int A[], int B[], int k, size_t sizeList){
     
    // temporal para almacenar los conteos, inicialmente a 0
    int C[k + 1] = {0};         

    // contando los valores repetidos
    for (int i = 0; i < sizeList; i++) {        
        C[A[i]] = C[A[i]] + 1;
    }

    // acumulando los repetidos para las posiciones
    for (int i = 1; i < (k + 1); i++)
    {
        /* code */
        C[i] = C[i] + C[i - 1];
    }    

    // ubicando los valores en sus respectivos indices
    for (int i = (sizeList - 1); i >= 0; i--)
    {
        // asignamos el valor en la posicion ordenada
        B[C[A[i]] - 1] = A[i];
        // descontamos en 1 al valor acumulado(posicion correcta)
        C[A[i]] = C[A[i]] - 1; 
    }
    return B;
}

/* Ejecucion */
int main()
{
	// Obtencion de los archivos generados
    vector<int> generate; 
    generate.push_back(100);
    generate.push_back(500);

    for(int i=1000 ; i < 10000 ; i+=1000) {
        generate.push_back(i);
    }
    for(int i=10000 ; i < 100001 ; i+=10000) {
        generate.push_back(i);
    }


    // Creacion del archivo para escritura de los items y el tiempo de ejecucion
    ofstream myfile;
    myfile.open ("countingSort_cpp_time.txt");

    for (int k = 0; k < generate.size(); k++)
    {
        // Obtencion de los items insertados a partir del tamanho ya obtenido anteriormente        
        ifstream infile2("../../1_preparacion_datos/generatedTestData/example_" + to_string(generate[k]) + ".txt");
        
        int arr[generate[k]]; //insertando el tamanho del array
        int i=0;
        string line;
        while (getline(infile2, line))
        {
            istringstream iss(line);
            int a;
            if (!(iss >> a)) { break; }			
            arr[i] = a;
            i++;
        }
		
        // solo para tener el numero de array, igualmente el valor ya es conocido
        int n = sizeof(arr) / sizeof(arr[0]);		     
		int B[generate[k]]={0};
		int maxValue = maxValueInArray(arr, n);	  		
			  
        // declaracion del tiempo de ejecucion del algoritmo
        clock_t start, end;
        start = clock();    
        counting_sort(arr,B,maxValue,generate[k]);
        end = clock();
		
        // tiempo de ejecucion obtenido
        double time_taken = double(end - start) / double(CLOCKS_PER_SEC); 
    
        cout << "Se ordeno el array de tamanho " + to_string(n) + " :";    
        cout << "\n Tiempo tomado por el programa es: " << fixed
             << time_taken << setprecision(8);
        cout << " segundos " << endl;

        //escribiendo en el archivo resultado
        myfile << generate[k] << " "; 
        myfile << fixed << time_taken;    
        myfile << "\n";
    }

    myfile.close();
}

File: Anexo_InformeLatex/EDA_Practica1/code/counting.py

# Algoritmo de ordenamiento - COUNTING SORT

import numpy as np
import sys
import time
import os
from tqdm import tqdm


# retorna el maximo valor de un arreglo de numeros
def maxValueInArray(lista, sizeList):
	
	max = 0
	# inicializando los contadores a 0
	for x in range(sizeList):
		max = max if max >= lista[x] else lista[x] 
	return max

# a = arreglo no ordenado, k = valor maximo del arreglo
def counting_sort(a, k):
	# temporal para acumular valores repetidos
	c = []
	         
	# arreglo ordenado para retornar, mismo tamanho del arreglo principal.
	b = [0] * len(a)
	
	# inicializando los contadores a 0
	for x in range(k + 1):
		c.append(0)    
	
	# contando los valores repetidos
	for x in range(len(a)):
		c[a[x]] = c[a[x]] + 1
	
	# acumulando los repetidos para las posiciones(indices del arreglo)
	for x in range(1, len(c)):
		c[x] = c[x] + c[x-1]	

	# ubicando los valores en sus respectivos indices
	for x in range(len(a) - 1, -1,-1):
		# asignamos el valor en la posicion ordenada
		b[c[a[x]] - 1] = a[x]
		# descontamos en 1 al valor acumulado(posicion correcta)
		c[a[x]] = c[a[x]] - 1
	return b


#Ejecucion

# Obtencion de los archivos generados
nameFiles_idx = np.concatenate((np.arange(100,501,400),np.arange(1000,10000,1000),np.arange(10000,100001,10000)),axis=None)

# Creacion del archivo para escritura de los items y el tiempo de ejecucion
file = open('countingSort_python_time.txt', "w")
dicPaths = {}
for namefile in nameFiles_idx:
  # para obtencion de items   
  dicPaths.update({namefile:'../../1_preparacion_datos/generatedTestData/example_'+str(namefile)+".txt"})  

  arr=[]
  with open(dicPaths[namefile], "r") as f:
    for line in f:
        arr += [int(line.strip())]

  # declaracion del tiempo de ejecucion del algoritmo
  start = time.process_time()  
  counting_sort(arr, maxValueInArray(arr, len(arr)))
  result = time.process_time() - start
  
  print("Se ordeno el array de tamanho " + str(len(arr)) + ": Tiempo tomado por el programa es... " + str(time.process_time() - start) + " segundos")

  # escribiendo en el archivo resultado
  file.write(str(namefile) + ' ' + str(result) +"\n")

file.close()